### ISSUE TITLE
add a timestamp

### ISSUE THREAD

**Issue by tobwen** [kellyjonbrazil/jc#265]
Would it be possible to add a switch to add a timestamp to each line? Many shell tools (like `free`) don't output a timestamp by default, but that way they could be used for logging.

**Comment by kellyjonbrazil** [comment]
Hi there - thanks for the suggestion. That's an interesting idea. Are you thinking the timestamp would be an additional key/value pair in the JSON output? Something like:
```
$ free | jc --free -p
    [
      {
        "_jc_timestamp": 12345678,
        "type": "Mem",
        "total": 3861340,
        "used": 220508,
        "free": 3381972,
        "shared": 11800,
        "buff_cache": 258860,
        "available": 3397784
      },
      {
        "_jc_timestamp": 12345678,
        "type": "Swap",
        "total": 2097148,
        "used": 0,
        "free": 2097148
      }
    ]
```
Might be able to do something like that. I'll have to look into it some more.

**Comment by tobwen** [comment]
On the one hand, adding it to each object makes it highly redundant, but on the other way, it's much easier to parse using most JSON tools.

**Comment by kellyjonbrazil** [comment]
I think the simplest method of doing this would be to keep the timestamp code in the `cli.py` module so I won't need to touch all of the parser modules.  That means that I can create timestamp objects like this for standard parser output:
```
[
  {"normal_output: "data"},
  {"normal_output: "data"},
  {"_jc_meta": {
    "timestamp": 12345
    }
  }
]
```
and
```
{
 "normal_output1": "data",
 "normal_output2": "data",
 "_jc_meta": {
    "timestamp": 12345
  }
}
```
and for streaming parsers, there would be a timestamp per output object:
```
{
 "normal_output1": "data",
 "normal_output2": "data",
 "_jc_meta": {
    "timestamp": 12345
  }
}
{
 "normal_output1": "data",
 "normal_output2": "data",
 "_jc_meta": {
    "timestamp": 12345
  }
}
```
If we go this route, then we just need to figure out which timestamp we are interested in. The start time or end time? Or maybe both. I could record the start time and then add it to the object along with the end time.

Also, do we want epoch timestamps or ISO datetime strings. Or both? The object could look something like:
"_jc_meta": {
  "start_timestamp": 12345,
  "end_timestamp": 12346,
  "start_iso": "2022-07-20T14:52:45Z",
  "end_iso": "2022-07-20T14:52:46Z"
}

I could do something like debug and quiet, where a single `-t` would give you just the start or end time, `-tt` would give you both, and `-ttt` would add on the ISO strings. Lots of options, just need to narrow down what makes sense.

**Comment by kellyjonbrazil** [comment]
I have this working in the `dev` branch. Here are some examples:

Standard parser that outputs a JSON object:
```
$ ping -c 2 1.1.1.1 | jc --ping -tp
{
  "destination_ip": "1.1.1.1",
  "data_bytes": 56,
  "pattern": null,
  "destination": "1.1.1.1",
  "packets_transmitted": 2,
  "packets_received": 2,
  "packet_loss_percent": 0.0,
  "duplicates": 0,
  "round_trip_ms_min": 26.68,
  "round_trip_ms_avg": 29.306,
  "round_trip_ms_max": 31.932,
  "round_trip_ms_stddev": 2.626,
  "responses": [
    {
      "type": "reply",
      "bytes": 64,
      "response_ip": "1.1.1.1",
      "icmp_seq": 0,
      "ttl": 56,
      "time_ms": 31.932,
      "duplicate": false
    },
    {
      "type": "reply",
      "bytes": 64,
      "response_ip": "1.1.1.1",
      "icmp_seq": 1,
      "ttl": 56,
      "time_ms": 26.68,
      "duplicate": false
    }
  ],
  "_jc_meta": {
    "timestamp": 1659624406.576528
  }
}
```

Streaming parser (each emitted JSON object has its own updated timestamp):
```
$ ping -c 2 1.1.1.1 | jc --ping-s -tp
{
  "type": "reply",
  "destination_ip": "1.1.1.1",
  "sent_bytes": 56,
  "pattern": null,
  "response_bytes": 64,
  "response_ip": "1.1.1.1",
  "icmp_seq": 0,
  "ttl": 56,
  "time_ms": 31.392,
  "_jc_meta": {
    "timestamp": 1659624494.985129
  }
}
{
  "type": "reply",
  "destination_ip": "1.1.1.1",
  "sent_bytes": 56,
  "pattern": null,
  "response_bytes": 64,
  "response_ip": "1.1.1.1",
  "icmp_seq": 1,
  "ttl": 56,
  "time_ms": 33.881,
  "_jc_meta": {
    "timestamp": 1659624495.865199
  }
}
{
  "type": "summary",
  "destination_ip": "1.1.1.1",
  "sent_bytes": 56,
  "pattern": null,
  "packets_transmitted": 2,
  "packets_received": 2,
  "packet_loss_percent": 0.0,
  "duplicates": 0,
  "round_trip_ms_min": 31.392,
  "round_trip_ms_avg": 32.636,
  "round_trip_ms_max": 33.881,
  "round_trip_ms_stddev": 1.245,
  "_jc_meta": {
    "timestamp": 1659624495.867864
  }
}
```

Standard parser that emits a JSON array of objects. (each object shares the same timestamp):
```
% arp -a | jc --arp -tp
[
  {
    "name": "kellys-mbp.attlocal.net",
    "address": "192.168.1.72",
    "hwtype": "ethernet",
    "hwaddress": "f0:18:98:3:d8:39",
    "iface": "en0",
    "permanent": false,
    "_jc_meta": {
      "timestamp": 1659624693.495372
    }
  },
  {
    "name": "c6180.attlocal.net",
    "address": "192.168.1.74",
    "hwtype": "ethernet",
    "hwaddress": "9c:b6:54:5a:5a:7c",
    "iface": "en0",
    "permanent": false,
    "_jc_meta": {
      "timestamp": 1659624693.495372
    }
  }
]
```

**Comment by kellyjonbrazil** [comment]
Here is an example of adding the timestamp to an existing `_jc_meta` field in a streaming parser:
```
% cat ping-ip-O-unparsedlines.out| jc --ping-s -tpqq
{
  "type": "reply",
  "destination_ip": "127.0.0.1",
  "sent_bytes": 56,
  "pattern": null,
  "timestamp": null,
  "response_bytes": 64,
  "response_ip": "127.0.0.1",
  "icmp_seq": 1,
  "ttl": 64,
  "time_ms": 0.038,
  "duplicate": false,
  "_jc_meta": {
    "success": true,
    "timestamp": 1659625316.273184
  }
}
{
  "type": "reply",
  "destination_ip": "127.0.0.1",
  "sent_bytes": 56,
  "pattern": null,
  "timestamp": null,
  "response_bytes": 64,
  "response_ip": "127.0.0.1",
  "icmp_seq": 2,
  "ttl": 64,
  "time_ms": 0.043,
  "duplicate": false,
  "_jc_meta": {
    "success": true,
    "timestamp": 1659625316.277439
  }
}
{
  "type": "reply",
  "destination_ip": "127.0.0.1",
  "sent_bytes": 56,
  "pattern": null,
  "timestamp": null,
  "response_bytes": 64,
  "response_ip": "127.0.0.1",
  "icmp_seq": 3,
  "ttl": 64,
  "time_ms": 0.044,
  "duplicate": false,
  "_jc_meta": {
    "success": true,
    "timestamp": 1659625316.278308
  }
}
{
  "_jc_meta": {
    "success": false,
    "error": "IndexError: list index out of range",
    "line": "64 bytes from 127.0.0.1: error - weird error",
    "timestamp": 1659625316.279121
  }
}
```

**Comment by kellyjonbrazil** [comment]
Released in v1.21.0 available via pip. Binaries coming soon.

A timestamp can be added by using the `--meta-out` or `-M` option. The metadata will look like this:
```
% jc --meta-out --pretty ping -c3 192.168.1.252
{
  "destination_ip": "192.168.1.252",
  "data_bytes": 56,
  "pattern": null,
  "destination": "192.168.1.252",
  "packets_transmitted": 3,
  "packets_received": 0,
  "packet_loss_percent": 100.0,
  "duplicates": 0,
  "responses": [
    {
      "type": "timeout",
      "icmp_seq": 0,
      "duplicate": false
    },
    {
      "type": "timeout",
      "icmp_seq": 1,
      "duplicate": false
    }
  ],
  "_jc_meta": {
    "parser": "ping",
    "timestamp": 1661115928.89416,
    "magic_command": [
      "ping",
      "-c3",
      "192.168.1.252"
    ],
    "magic_command_exit": 2
  }
}
```
